safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,170 @@
+// limitations under the License.
+
+package csp
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// Directive is the name of a single CSP directive.
+type Directive string
+
+const (
+	DirectiveScriptSrc Directive = "script-src"
+	DirectiveStyleSrc            = "style-src"
+	DirectiveObjectSrc           = "object-src"
+	DirectiveBaseURI             = "base-uri"
+	DirectiveReportURI           = "report-uri"
+)
+
+const (
+	ValueHTTPS         = "https:"
+	ValueHTTP          = "http:"
+	ValueUnsafeEval    = "'unsafe-eval'"
+	ValueUnsafeInline  = "'unsafe-inline'"
+	ValueNone          = "'none'"
+	ValueStrictDynamic = "'strict-dynamic'"
+)
+
+// PolicyDirective contains a single CSP directive.
+type PolicyDirective struct {
+	Directive Directive
+	Values    []string
+	AddNonce  bool
+}
+
+// nonceSize is the size of the nonces in bytes.
+const nonceSize = 8
+
+func generateNonce(readRand func([]byte) (int, error)) string {
+	if readRand == nil {
+		readRand = rand.Read
+	}
+	b := make([]byte, nonceSize)
+	_, err := readRand(b)
+	if err != nil {
+		// TODO: handle this better, what should happen here?
+		panic(err)
+	}
+	return base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy, containing many directives.
+type Policy struct {
+	Directives []*PolicyDirective
+	// readRand is used for dependency injection in tests.
+	readRand func([]byte) (int, error)
+}
+
+// NewPolicy creates a new strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// TODO: maybe reportURI should be safehttp.URL?
+func NewPolicy(reportURI string) *Policy {
+	return &Policy{
+		Directives: []*PolicyDirective{
+			{Directive: DirectiveObjectSrc, Values: []string{ValueNone}, AddNonce: false},
+			{
+				Directive: DirectiveScriptSrc,
+				Values: []string{
+					ValueUnsafeInline,
+					ValueUnsafeEval,
+					ValueStrictDynamic,
+					ValueHTTPS,
+					ValueHTTP,
+				},
+				AddNonce: true,
+			},
+			{Directive: DirectiveBaseURI, Values: []string{ValueNone}, AddNonce: false},
+			{Directive: DirectiveReportURI, Values: []string{reportURI}, AddNonce: false},
+		},
+	}
+}
+
+// Serialize serializes this policy for use in a Content-Security-Policy header
+// or in a Content-Security-Policy-Report-Only header. The nonces generated for
+// each directive are also returned.
+func (p Policy) Serialize() (csp string, nonces map[Directive]string) {
+	nonces = make(map[Directive]string)
+	values := make([]string, 0, len(p.Directives))
+
+	for _, d := range p.Directives {
+		var b strings.Builder
+		b.WriteString(string(d.Directive))
+
+		if d.AddNonce {
+			n := generateNonce(p.readRand)
+			b.WriteString(fmt.Sprintf(" 'nonce-%s'", n))
+			nonces[d.Directive] = n
+		}
+
+		b.WriteString(" ")
+		b.WriteString(strings.Join(d.Values, " "))
+
+		values = append(values, b.String())
+	}
+
+	return strings.Join(values, "; "), nonces
+}
```

Comment:
I would make serialize take a context and if there is already a nonce there use it, otherwise add a new one and return the new context. Let's not expose a map of nonces. Also, nonce should be only one per response.

If you plan to still return the nonce please unexport this method and provide one that behaves as I said if someone wants to use the plugin without the framework.